### ISSUE TITLE
Cannot update to nix 2.19.1

### ISSUE THREAD

**Issue by dwt** [DeterminateSystems/nix-installer#744]
From the [NixOS](https://nixos.org/download#nix-install-macos) homepage it seems that version 2.19.1 is already released. However, invoking the documented way to self update nix with `sudo -i nix upgrade-nix` doesn't seem to actually update to that version, instead staying on the old version 2.18.1.

This seems like a bug, but as a beginner with nix this might surely be my misunderstanding of how nix releases work, so I'd love to be enlightened.

**Comment by grahamc** [comment]
Hey @dwt, good question. By default, Nix's self-update mechanism gets the latest version available in Nixpkgs, which is not very predictable or "timely". We haven't updated to Nix 2.19.1 yet because of some UX changes regarding flake lock updating, and a regression that was identified in 2.19.0. We'll be vetting 2.19 and getting that out soon.

We're going to do some work to make the upgrade-nix behavior more predictable and understandable, though, and also have it track the latest version of Nix that Determinate Nix Installer uses.

It sounds like the biggest issue here is unclear communication about what is in use and why. Is that right? That should get cleared up when we have it track the latest we install. Does this make sense to you?

**Comment by afh** [comment]
I'd find clearer communication about what is in use and why very helpful.

**Comment by dwt** [comment]
That sounds good? I'm not quite sure I understand what you say, but to me it sounds like there was an active decision not to make `nix upgrade-nix` upgrade to the latest version of nix, but instead wait for some bugfixes?

In that case, a clearer output why an old version is selected by `nix upgrade-nix` would surely help, especially if that points to the ticket where those regressions and the reasoning is tracked? That would make this completely transparent and easy to track when the update is ready.

**Comment by grahamc** [comment]
Got it. Thanks.

To share more context:

* Nix looks at Nixpkgs to find the release of Nix to upgrade to. It does this by fetching https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/nix-fallback-paths.nix.
* The Nix team is not responsible for when those paths get updated, the Nixpkgs maintenance team is. This makes for a confusing situation...

We did make an active decision to not upgrade to Nix 2.19, because of https://github.com/NixOS/nix/issues/9414.

That has been fixed in 2.19.1, but there was another problem. 2.19 removed support for updating flakes via `nix flake lock --update-input`, and renamed it to `nix flake update`. This is a good improvement, but the change didn't help users understand the migration. The Determinate Nix promise on flake stability is on locked expressions, and not necessarily the CLI. However, we do still strive to make changes to the CLI as painless as possible. Because of that, we're tentatively waiting on https://github.com/NixOS/nix/pull/9449 to make that migration easier.

Once 2.19 is out in the wild through our installer, we'll be in a better place to clearly communicate the status. We can also make install.determinate.systems have status information about Nix releases and why we are shipping what we are. I'll put in some tickets on this and get the ball rolling.

Thank you for the feedback!

**Comment by dwt** [comment]
Thanks a lot for the pointer to the tickets, and if I understand you right, you are planning to communicate that information from the shell. I love that.

**Comment by grahamc** [comment]
I hope we will be able to and we'll try, but we may not be able to. We might have to send some patches upstream to make that possible 👍. Can you give an example of what you'd like that to look like?

**Comment by dwt** [comment]
```shell
❯ sudo -i nix upgrade-nix
Password:
replacing old 'nix-2.18.1'
installing 'nix-2.18.1'
Note: nix-2.19.1 is already released, but due to bug https://github.com/... we've decided not to install it with the determinate nix installer yet. Follow that bug to stay up to date when the Update will happen.

~ 4s 
❯ 
```

Something like this perhaps?

**Comment by a-h** [comment]
I ran into this today. I have a Nix flake that builds fine on my machine, but when a colleague builds the same flake, it fails with:

```
error: hash mismatch importing path '/nix/store/c65c22j7fcirs5njfjv9ls0w993hx0ia-gcc-wrapper-13.2.0-man';
         specified: sha256:0iraz5v8vmbsixm0hmrra0g4yr8avbsywxf5fa0sp9xpjgj1374b
         got:       sha256:1jj7rnild9l552xn5i46rlbr9j50i0rcdi1vxvg7p915wv5khl8p
```

It must be a bug in Nix, because "reproducible builds". I narrowed it down to the fact that he recently installed Nix using the Determinate Systems installer, and got Nix 2.19, but I had Nix 2.18... (https://discourse.nixos.org/t/error-hash-mismatch-importing-path/26353)

So, I thought I'd try to upgrade to Nix 2.20.1 using the Determinate Systems installer to see if the bug was fixed in a more recent version of the Nix CLI (this issue looked promising, and was released in 2.20.1 https://github.com/NixOS/nix/issues/9143), but couldn't find a way to upgrade. I was hoping that a command like this would work in the installer.

```
./nix-installer-x86_64-linux upgrade --nix-package-url https://releases.nixos.org/nix/nix-2.20.1/nix-2.20.1-x86_64-linux.tar.xz
```

Unfortunately not, and as far as I can see, the Nix upgrade won't take me to that version, so I uninstalled Nix with the Determinate Systems installer, installed a later version with `./nix-installer-x86_64-linux upgrade --nix-package-url https://releases.nixos.org/nix/nix-2.20.1/nix-2.20.1-x86_64-linux.tar.xz` but then had to reinstall all of my `home-manager` config which took up time as it downloaded / built everything again.

I think that an upgrade option in the installer that just upgrades the Nix CLI instead of having to destroy and recreate the `/nix` store path would be handy.

**Comment by develop7** [comment]
> I think that an upgrade option in the installer that just upgrades the Nix CLI instead of having to destroy and recreate the `/nix` store path would be handy.

I've just upgraded nix to latest version by running 
```sh
$ nix upgrade-nix --nix-store-paths-url https://raw.githubusercontent.com/DeterminateSystems/nix-upgrade/main/versions.nix
```
then adding 
```nix
upgrade-nix-store-path-url = https://install.determinate.systems/nix-upgrade/stable/universal
```
to the `/etc/nix/nix.conf`

**Comment by ilyagr** [comment]
This worked well for me:

      nix upgrade-nix --nix-store-paths-url https://releases.nixos.org/nix/nix-2.19.4/fallback-paths.nix


Unfortunately, I don't know how to find out automatically what the latest version of nix 2.19 is. Don't use 2.19.3; it has some sort of a security vulnerability. `curl` succeeds on every 2.19.X version.

      nix upgrade-nix --nix-store-paths-url https://releases.nixos.org/nix/nix-2.20.2/fallback-paths.nix

should also work.

**Comment by bestlem** [comment]
On macOs the last suggestion gives
`error: Nix on NixOS must be upgraded via 'nixos-rebuild'`

**Comment by tedbyron** [comment]
> On macOs the last suggestion gives `error: Nix on NixOS must be upgraded via 'nixos-rebuild'`

@bestlem Are you using nix-darwin? If so, the first nix in your $PATH is configured with `nix.package`. I am using `nix.package = unstable.nixVersions.nix_2_20;` where unstable is nixpkgs-unstable

**Comment by cole-h** [comment]
> On macOs the last suggestion gives `error: Nix on NixOS must be upgraded via 'nixos-rebuild'`

If you're using `nix-darwin`, this will trigger that message (this message just checks for the existence of `/run/current-system` IIRC, which NixOS creates, and `nix-darwin` was inspired by).

You'll likely want to use `nix-darwin` to manage your Nix version (or uninstall `nix-darwin` before trying to run `nix upgrade-nix`).

---

As of https://github.com/DeterminateSystems/nix-installer/pull/772 (included and set up by `nix-installer` v0.16.0 and later), you should now be able to upgrade Nix with just `nix upgrade-nix` (properly elevated as necessary).

If you had installed with an older version, you can use `nix upgrade-nix --nix-store-paths-url https://install.determinate.systems/nix-upgrade/stable/universal` or add `upgrade-nix-store-path-url = https://install.determinate.systems/nix-upgrade/stable/universal` to your `nix.conf`.
